### PR TITLE
KAFKA-17776: Capability for connection disconnect listeners in socket server

### PR DESCRIPTION
**Pull request by apoorvmittal10**
As of now, there isn't any way that resources can be cleaned in Kafka when client disconnects the connections. Generally time bound clean up happens which keep the resources occupied.

This PR provides capability to attach listeners in Socket Server which can react to client connection disconnections. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/network/SocketServer.scala [hunk 1]
```diff
@@ -1220,6 +1227,8 @@ private[kafka] class Processor(
         inflightResponses.remove(connectionId).foreach(updateRequestMetrics)
         // the channel has been closed by the selector but the quotas still need to be updated
         connectionQuotas.dec(listenerName, InetAddress.getByName(remoteHost))
+        // Call listeners to notify for closed connection.
+        connectionDisconnectListeners.foreach(listener => listener.onDisconnect(connectionId))
```
Comment (chia7712) [inline core/src/main/scala/kafka/network/SocketServer.scala]: Should we also handle the other path that doesn't update disconnected? connectionQuotas already manages both paths.

https://github.com/apache/kafka/blob/76a1af984b39d9890fe26954aff36bb1a321af77/core/src/main/scala/kafka/network/SocketServer.scala#L1244

File: server-common/src/main/java/org/apache/kafka/server/network/ConnectionDisconnectListener.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ */
+public interface ConnectionDisconnectListener {
+
+    /**
+     * Invoked when a connection is disconnected.
+     *
+     * @param connectionId The connection id as defined in {@link org.apache.kafka.common.requests.RequestContext}.
+     *                     This id is unique for each connection.
+     */
+    void onDisconnect(String connectionId);
```
Comment (chia7712) [inline server-common/src/main/java/org/apache/kafka/server/network/ConnectionDisconnectListener.java:35]: Could you please add comment to remind developers that this method will block processor so it should be fast.

File: core/src/main/scala/kafka/network/SocketServer.scala (hunk 1 above)
Comment (chia7712) [reply core/src/main/scala/kafka/network/SocketServer.scala]: Additionally, should we swallow the exception in the loop to ensure all listeners are executed?

**Review by chia7712 — COMMENTED** [review]
@apoorvmittal10 thanks for this patch.

File: core/src/main/scala/kafka/network/SocketServer.scala (hunk 1 above)
Comment (apoorvmittal10) [reply core/src/main/scala/kafka/network/SocketServer.scala]: Thanks for pointing out, agree. Done.

File: server-common/src/main/java/org/apache/kafka/server/network/ConnectionDisconnectListener.java (hunk 2 above)
Comment (apoorvmittal10) [reply server-common/src/main/java/org/apache/kafka/server/network/ConnectionDisconnectListener.java:35]: Added, done.

**Comment by apoorvmittal10** [comment]
@chia7712 I have addressed comments and build passed.

File: core/src/main/scala/kafka/network/SocketServer.scala [hunk 3]
```diff
@@ -1248,6 +1257,8 @@ private[kafka] class Processor(
       if (address != null)
         connectionQuotas.dec(listenerName, address)
       selector.close(connectionId)
+      // Call listeners to notify for closed connection.
+      connectionDisconnectListeners.foreach(listener => CoreUtils.swallow(() -> listener.onDisconnect(connectionId), this, Level.ERROR))
```
Comment (chia7712) [inline core/src/main/scala/kafka/network/SocketServer.scala:1261]: The connection ID is not a UUID, so the connection ID generated by processor_a could be the same as the one generated by processor_b. However, in the follow-up PR #17474, all processors' connection IDs are stored in a single clientConnectionIdMap. How do we handle connection ID conflicts in that case? This question may not be directly related to this PR, but it would be great to discuss it early.

Comment (apoorvmittal10) [reply core/src/main/scala/kafka/network/SocketServer.scala:1261]: Yeah connectionId is not UUID but they are unique. The format they have is "<kafka_host_IP>-<kafka_port>-<remote_host_ip>-<remote_port>-<montonically_increasing_numer>". If this won't be unique then we would have bigger issues as Selector.java also keeps a map of connectionId to KafkaChannel which mutes or disconnect channels based on this connectionId.

Comment (chia7712) [reply core/src/main/scala/kafka/network/SocketServer.scala:1261]: > The format they have is "<kafka_host_IP>-<kafka_port>-<remote_host_ip>-<remote_port>-<montonically_increasing_numer>". If this won't be unique then we would have bigger issues as Selector.java also keeps a map of connectionId to KafkaChannel which mutes or disconnect channels based on this connectionId.

Yes, but the ID is only unique within the scope of a single process. Since `monotonically_increasing_number` is a local variable of each processor, it's possible for two processors to generate the same ID. Additionally, processors don't share the same selector, so the ID doesn't need to be globally unique.

My concern was the duplicate id will corrupt the cache of (id -> UUID)

Comment (apoorvmittal10) [reply core/src/main/scala/kafka/network/SocketServer.scala:1261]: That's a good point. I digged more into the SocketServer code and seems to me that Processors are attached to Acceptors. There could be control plane or data plane acceptors which are instantiated as per server `listeners` configuration.
So that means, there could be a possibility that if: 2 different clients + on different acceptors + with same remote host and port + generates same counter then can have same connection id. Hence do you think that we should also have listener name and security protocol passed to disconnect listeners. This information is also available in RequestContext. Or listenerName alone will be sufficient as like `connectionQuotas`?

**Comment by apoorvmittal10** [comment]
@chia7712 I have added listener name and security protocol which defines the endpoint and acceptors in the disconnection method. Can you please re-review.

**Comment by apoorvmittal10** [comment]
Unrelated flaky test on Java 21, build passed on java 11.

File: core/src/main/scala/kafka/network/SocketServer.scala (hunk 3 above)
Comment (junrao) [reply core/src/main/scala/kafka/network/SocketServer.scala:1261]: Hmm, two different clients from the same remote host should have different remote ports, right?

Comment (junrao) [reply core/src/main/scala/kafka/network/SocketServer.scala:1261]: Ok. The reason that we added monotonically_increasing_number is that a client could reuse the same port after closing but before the server realizes it. So connectionId is unique within a Processor, but not necessarily unique across Processor. Perhaps we could use both connectionId and processor id. For processors under different listeners, they have different server port. So, it doesn't seem that we need to include the listener name.

Comment (apoorvmittal10) [reply core/src/main/scala/kafka/network/SocketServer.scala:1261]: @junrao You are right. 1 query I have, I missed that kafka port will also change with processor hence isn't connectionId is then unique across processors as well?

Comment (apoorvmittal10) [reply core/src/main/scala/kafka/network/SocketServer.scala:1261]: I have reverted the last commit of mine which added listener and security protocol. Also seems to me that processorId is also not required as server port will change with processor hence making connectionId unique.

Comment (chia7712) [reply core/src/main/scala/kafka/network/SocketServer.scala:1261]: > I missed that kafka port will also change with processor hence isn't connectionId is then unique across processors as well?

Please correct me if I'm wrong.

1. `<kafka_host_IP>-<kafka_port>`: The Processor handles connections accepted by the Acceptor, so sockets handled by different processors can have the same `<kafka_host_IP>-<kafka_port>`. This means that `<kafka_host_IP>-<kafka_port>` is not unique across processors.
2. `<remote_host_ip>-<remote_port>`: As @junrao mentioned, this part is not unique either.
3. `<monotonically_increasing_number>`: This is a local incremental variable, so it is also not unique.

In short, current implementation of `connectionId` is not unique across processors.

Comment (apoorvmittal10) [reply core/src/main/scala/kafka/network/SocketServer.scala:1261]: @chia7712 Make sense. There are 2 options:

1. Expose processor id in RequestContext so at other places, unique connection reference can be attached.
2. Add processor id in connection id i.e. `<kafka_host_IP>:<kafka_port>-<remote_host_ip>:<remote_port>-<monotonically_increasing_number>-<processor-id>`

@chia7712 @junrao I am inclined towards 2, do you see any issue in adding processor id in connection id creation?

Comment (chia7712) [reply core/src/main/scala/kafka/network/SocketServer.scala:1261]: > <kafka_host_IP>:<kafka_port>-<remote_host_ip>:<remote_port>-<monotonically_increasing_number>-<processor-id>

It seems this PR uses `<kafka_host_IP>:<kafka_port>-<remote_host_ip>:<remote_port>-<processor-id>-<monotonically_increasing_number>`, right?

Comment (apoorvmittal10) [reply core/src/main/scala/kafka/network/SocketServer.scala:1261]: Yeah, while implementing I thought to have processor id prior from `monotonically_increasing_number` as the `monotonically_increasing_number` belongs to processor id.

File: clients/src/main/java/org/apache/kafka/common/network/Selector.java [hunk 4]
```diff
@@ -237,15 +237,16 @@ public Selector(long connectionMaxIdleMS, int failedAuthenticationDelayMs, Metri
      * Generates a unique connection ID for the given socket.
      *
      * @param socket The socket for which the connection ID is to be generated.
+     * @param processorId The ID of the processor that will handle this connection.
```
Comment (junrao) [inline clients/src/main/java/org/apache/kafka/common/network/Selector.java]: The ID of the processor => The ID of the server processor

File: core/src/main/scala/kafka/network/SocketServer.scala [hunk 5]
```diff
@@ -929,17 +936,17 @@ private[kafka] class Processor(
...
+          ConnectionId(localHost, localPort, remoteHost, remotePort, Integer.parseInt(processorId), Integer.parseInt(index))
         }
       }
       case _ => None
     }
   }
 
-  private[network] case class ConnectionId(localHost: String, localPort: Int, remoteHost: String, remotePort: Int, index: Int) {
-    override def toString: String = s"$localHost:$localPort-$remoteHost:$remotePort-$index"
+  private[network] case class ConnectionId(localHost: String, localPort: Int, remoteHost: String, remotePort: Int, processorId: Int, index: Int) {
```
Comment (junrao) [inline core/src/main/scala/kafka/network/SocketServer.scala:948]: The connectionId logic is split between `SocketServer` and `Selector.generateConnectionId`. Perhaps we could consolidate them into a top level class ConnectionIdForServer, which can be constructed from either a string or a socket. It also has a `toString` method to get the string representation of the connectionId.

File: core/src/test/scala/unit/kafka/network/SocketServerTest.scala [hunk 6]
```diff
@@ -2050,6 +2053,25 @@ class SocketServerTest {
     }
   }
 
+  @Test
+  def testConnectionDisconnectListenerInvokedOnClose(): Unit = {
+    var listenerConnectionId: String = ""
```
Comment (junrao) [inline core/src/test/scala/unit/kafka/network/SocketServerTest.scala]: Should this be volatile since it's read and written by different threads?

**Review by junrao — COMMENTED** [review]
@apoorvmittal10 : Thanks for the updated PR. A few more comments.

File: clients/src/main/java/org/apache/kafka/common/network/Selector.java (hunk 4 above)
Comment (apoorvmittal10) [reply clients/src/main/java/org/apache/kafka/common/network/Selector.java]: Done.

File: core/src/test/scala/unit/kafka/network/SocketServerTest.scala (hunk 6 above)
Comment (apoorvmittal10) [reply core/src/test/scala/unit/kafka/network/SocketServerTest.scala]: Agree, done.

File: core/src/main/scala/kafka/network/SocketServer.scala (hunk 5 above)
Comment (apoorvmittal10) [reply core/src/main/scala/kafka/network/SocketServer.scala:948]: Yeah, agree. I started it and realized we also need to port BrokerEndpoint.scala and corresponding changes. Shall I take that in follow up PR?

Comment (apoorvmittal10) [reply core/src/main/scala/kafka/network/SocketServer.scala:948]: I have created following jira: https://issues.apache.org/jira/browse/KAFKA-17813. Working on it, will send the patch. It's a medium refactoring change with tests where BrokerEndpoint is moved as well to java.

**Comment by apoorvmittal10** [comment]
@junrao @chia7712 Thanks for reviewing, I have addressed comments and the build passed.

**Review by junrao — APPROVED** [review]
@apoorvmittal10 : Thanks for the updated PR. LGTM. 

@chia7712 : Do you have any more comments?

**Review by chia7712 — APPROVED** [review]
